pkg/apiserver/apiserver_test.go

```diff
@@ -3157,6 +3172,121 @@ func TestUpdateChecksAPIVersion(t *testing.T) {
+}
+
+func (obj *SimpleXGSubresource) GetObjectKind() unversioned.ObjectKind { return &obj.TypeMeta }
+
+type SimpleXGSubresourceRESTStorage struct {
+	item SimpleXGSubresource
+}
+
+func (storage *SimpleXGSubresourceRESTStorage) New() runtime.Object {
+	return &SimpleXGSubresource{}
+}
+
+func (storage *SimpleXGSubresourceRESTStorage) Get(ctx api.Context, id string) (runtime.Object, error) {
+	copied, err := api.Scheme.Copy(&storage.item)
+	if err != nil {
+		panic(err)
+	}
+	return copied, nil
+}
+
+func TestXGSubresource(t *testing.T) {
+	container := restful.NewContainer()
+	container.Router(restful.CurlyRouter{})
+	mux := container.ServeMux
+
+	itemID := "theID"
+	subresourceStorage := &SimpleXGSubresourceRESTStorage{
+		item: SimpleXGSubresource{
+			SubresourceInfo: "foo",
+		},
+	}
+	storage := map[string]rest.Storage{
+		"simple":           &SimpleRESTStorage{},
+		"simple/subsimple": subresourceStorage,
+	}
+
+	group := APIGroupVersion{
+		Storage: storage,
+
+		RequestInfoResolver: newTestRequestInfoResolver(),
+
+		Creater:   api.Scheme,
+		Convertor: api.Scheme,
+		Typer:     api.Scheme,
+		Linker:    selfLinker,
+		Mapper:    namespaceMapper,
+
+		ParameterCodec: api.ParameterCodec,
+
+		Admit:   admissionControl,
+		Context: requestContextMapper,
+
+		Root:                   "/" + prefix,
+		GroupVersion:           testGroupVersion,
+		OptionsExternalVersion: &testGroupVersion,
+		Serializer:             api.Codecs,
+
+		SubresourceGroupVersionKind: map[string]unversioned.GroupVersionKind{
+			"simple/subsimple": testGroup2Version.WithKind("SimpleXGSubresource"),
+		},
+	}
+
+	if err := (&group).InstallREST(container); err != nil {
+		panic(fmt.Sprintf("unable to install container %s: %v", group.GroupVersion, err))
+	}
+
+	ws := new(restful.WebService)
+	InstallSupport(mux, ws)
+	container.Add(ws)
+
+	handler := defaultAPIServer{mux, container}
+	server := httptest.NewServer(handler)
+	// TODO: Uncomment when fix #19254
+	// defer server.Close()
+
+	resp, err := http.Get(server.URL + "/" + prefix + "/" + testGroupVersion.Group + "/" + testGroupVersion.Version + "/namespaces/default/simple/" + itemID + "/subsimple")
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+	if resp.StatusCode != http.StatusOK {
+		t.Fatalf("unexpected response: %#v", resp)
+	}
+	var itemOut SimpleXGSubresource
+	body, err := extractBody(resp, &itemOut)
+	if err != nil {
+		t.Errorf("unexpected error: %v", err)
+	}
+
+	// Test if the returned object has the expected group, version and kind
+	// We are directly unmarshaling JSON here because TypeMeta cannot be decoded through the
+	// installed decoders. TypeMeta cannot be decoded because it is added to the ignored
+	// conversion type list in API scheme and hence cannot be converted from input type object
+	// to output type object. So it's values don't appear in the decoded output object.
+	decoder := json.NewDecoder(strings.NewReader(body))
```

Comment:
@liggitt We can decode a serialized API object into a destination, not just `TypeMeta`. In fact we can decode `TypeMeta` just fine, but the decoded values won't be copied to the destination object (`itemOut` here) and that is by design. It would be dangerous to overwrite `TypeMeta`. We could probably implement a converter to allow overwriting `TypeMeta` in tests, but I don't think it is worth spending time on it.

I think this is a reasonable workaround. Do you have any other alternative suggestions?